gobblin-core-base/src/main/java/gobblin/writer/AsyncWriterManager.java

```diff
@@ -0,0 +1,519 @@
+    this.asyncDataWriter = asyncDataWriter;
+    this.closer.register(asyncDataWriter);
+
+
+    this.writeCallback = new WriteCallback() {
+      @Override
+      public void onSuccess(WriteResponse writeResponse) {
+        recordsSuccess.mark();
+        log.debug("Received success", writeResponse.getStringResponse());
+      }
+
+      @Override
+      public void onFailure(Throwable throwable) {
+        recordsFailed.mark();
+      }
+    };
+  }
+
+  @Override
+  public void write(final D record)
+      throws IOException {
+    /**
+     * A few ways to achieve max concurrency.
+     * a) Keep a counter that increments and decrements.
+     * Problem: you have to spin lock on the counter on a write... may not be too bad.
+     * Dangers: If you don't decrement the counter on corner cases, you may have a false sense of how many outstanding
+     * requests you have.
+     * b) Use a bounded queue to store call attempts, write the returned futures into it, remove the future on callback success.
+     * Problem: More state to keep in memory.
+     * How do you deal with retries.. how do you ensure there is no starvation on retries?
+     *
+     * Doing a) for now.
+     */
+
+    maybeThrow();
+    int spinNum = 0;
+    while (recordsIn.getCount() - (recordsSuccess.getCount() + recordsFailed.getCount()) > maxOutstandingWrites)
+    {
+      // spin until we have space to write. Sleep 1 millis every spin
+      try {
+        Thread.sleep(1);
+        ++spinNum;
+        if (spinNum % 50 == 0) {
+          log.info("Spinning due to pending writes, in = " + recordsIn.getCount() +
+              ", success = " + recordsSuccess.getCount() + ", failed = " + recordsFailed.getCount() +
+              ", maxOutstandingWrites = " + maxOutstandingWrites);
+        }
+      } catch (InterruptedException e) {
+        Throwables.propagate(e);
+      }
+    }
+    recordsIn.mark();
+    if (this.retriesEnabled) {
+      attemptWrite(new Attempt(record, 1, null));
+    } else {
+      recordsAttempted.mark();
+      this.asyncDataWriter.write(record, writeCallback);
+    }
+  }
+
+  /**
+   * Checks if the current operating metrics would imply that
+   * we're out of SLA on failures permitted
+   * @return true if any failure would be fatal
+   *
+   * TODO: Add windowed stats to test for x% failures in y time window
+   *
+   */
+  private boolean isFailureFatal()
+  {
+    return (failureAllowance == 0.0);
+  }
+
+
+  private void makeNextWriteThrow(Throwable t)
+  {
+    cachedWriteException = t;
+  }
+
+  private void maybeThrow() throws IOException {
+    if (cachedWriteException != null)
+    {
+      throw new IOException("Irrecoverable failure on async write", cachedWriteException);
+    }
+  }
+
+  private void attemptWrite(final Attempt attempt)
+  {
+    final long startTime = System.nanoTime();
+    this.recordsAttempted.mark();
+    this.asyncDataWriter.write(attempt.record, new WriteCallback() {
+
```

Comment:
If there are no more calls to write() then we don't have any way to indicate to the calling thread (the Fork) that we need to throw an exception. 

The next place where this can happen is in the commit() where SLA-s are getting checked. 
What I can do is wrap the cached write exception at that point and throw it with the IOException that gets thrown.